### PR TITLE
Support `__cmp__` for Python 2

### PR DESCRIPTION
**Pull request by rowillia**
This fixes https://github.com/python/mypy/issues/1165

### CODE REVIEW

File: test-data/unit/check-expressions.test [hunk 1]
```diff
@@ -410,6 +410,53 @@ class B:
 main:3: error: Incompatible types in assignment (expression has type "bool", variable has type "A")
 main:4: error: Incompatible types in assignment (expression has type "bool", variable has type "A")
 
+[case testCmp_python2]
+
+class bool: pass
```
Comment (gvanrossum) [inline test-data/unit/check-expressions.test]: Why not use `[builtins fixtures/bool.py]` instead?

File: test-data/unit/check-expressions.test [hunk 2]
```diff
@@ -410,6 +410,53 @@ class B:
...
+    def __cmp__(self, o):
+        # type: ('B') -> bool
+        pass
+    def __le__(self, o):
+        # type: ('C') -> bool
+        pass
+class C:
+    pass
+[out]
+main:5: error: Argument 1 to "__cmp__" of "A" has incompatible type "A"; expected "B"
```
Comment (gvanrossum) [inline test-data/unit/check-expressions.test]: Can you switch to the `# E: ...` convention? It makes it easier to see what lines are supposed to produce which errors.

File: test-data/unit/check-expressions.test (hunk 1 above)
Comment (gvanrossum) [inline test-data/unit/check-expressions.test]: (More precisely, `[builtins_py2 fixtures/bool.py]`.)

Comment (rowillia) [inline test-data/unit/check-expressions.test]: Thanks!  I was missing the `_py2` syntax 😄

File: test-data/unit/check-expressions.test (hunk 2 above)
Comment (rowillia) [inline test-data/unit/check-expressions.test]: Done

**Comment by rowillia** [comment]
Thanks for the feedback!  Digging into the tests a little more I realized that the matching wasn't quite robust enough.  I added in falling back to `__cmp__` on the after finding a `__rX` method fails, and checking for `__cmp__` both on the LHS and on the RHS.

File: mypy/checkexpr.py [hunk 3]
```diff
@@ -1061,15 +1061,54 @@ def check_op(self, method: str, base_type: Type, arg: Node,
             # method of the other operand instead.
             rmethod = self.get_reverse_op_method(method)
             arg_type = self.accept(arg)
+            base_arg_node = TempNode(base_type)
+            error_results = []  # type: List[Tuple[Tuple[Type, Type], MessageBuilder]]
```
Comment (gvanrossum) [inline mypy/checkexpr.py]: Hm... The logic involving this variable is confusing. If you're just going to report the first error, then why have a list? And should we really report the first error, not the last one?

(I do understand some of the logic -- if there's an error you still have to try other things which may succeed. But it's a lot of code.)

(Also, you can probably save on line breaks by setting your limit to 99 chars, which is what we use -- we've given up on 80 for mypy.)

Comment (rowillia) [inline mypy/checkexpr.py]: @gvanrossum I can make this simpler for now.  I was attempting to preserve the current behavior for now while in a follow up PR constructing a better error message when we fail detailing all of the methods that were considered and why they were not usable.  

I find the error messages from the current implementation to be somewhat confusing.  For example:

```py
class A(object):
  def __lt__(self, other: 'B'):
    pass
class B(object):
  pass
class C(object):
  def __gt__(self, other: 'B'):
    pass
A() < C() # E: Unsupported operand types for > ("C" and "A")
```

In this case I would expect to get an error message about `<`, not `>`


RE: 80 chars 😆

**Comment by rowillia** [comment]
@gvanrossum Anything else I can do here for this PR?

**Comment by gvanrossum** [comment]
Thanks for the ping -- I may have to sleep on this just because it's part of mypy's code that I'm not super familiar with, and we encountered a few surprises during development of the PR suggesting it's not trivial code. Maybe @JukkaL can review it in the next few days? Otherwise I promise I'll take a look next week (I'm taking the rest of this week off).

**Comment by rowillia** [comment]
Great.  Thanks @gvanrossum and thanks for removing `__cmp__` from Python 3 to reduce complexity around comparisons 😄 .

@JukkaL let me know if you have any questions here.  Most of the complexity comes from the evaluation precedence.  For example:

```py
A < B

#  Looks for a match in the following order:

A.__lt__(B)
B.__gt__(B)
A.__cmp__(B)
B.__cmp__(B)
```

File: mypy/checkexpr.py [hunk 4]
```diff
@@ -1061,15 +1061,51 @@ def check_op(self, method: str, base_type: Type, arg: Node,
             # method of the other operand instead.
             rmethod = self.get_reverse_op_method(method)
             arg_type = self.accept(arg)
+            base_arg_node = TempNode(base_type)
+            user_error = None  # type: Tuple[Tuple[Type, Type], MessageBuilder]
```
Comment (JukkaL) [inline mypy/checkexpr.py]: The type should be `Optional[Tuple[...]]`? (Not that it matters to mypy right now, but soon it hopefully will matter.)

Comment (JukkaL) [inline mypy/checkexpr.py]: Also, the name of the variable is a little confusing. Maybe the name could be more informative? Alternatively, add a comment.

File: mypy/checkexpr.py [hunk 5]
```diff
@@ -1061,15 +1061,51 @@ def check_op(self, method: str, base_type: Type, arg: Node,
...
             if self.has_member(arg_type, rmethod):
-                method_type = self.analyze_external_member_access(
-                    rmethod, arg_type, context)
-                temp = TempNode(base_type)
-                return self.check_call(method_type, [temp], [nodes.ARG_POS],
-                                       context)
+                local_errors = self.msg.copy()
```
Comment (JukkaL) [inline mypy/checkexpr.py]: This code gets repeated almost identically three times. Maybe define a nested function that avoids the repetition? It would probably also make the function shorter -- it's getting pretty long and somewhat hard to follow.

**Comment by rowillia** [comment]
Thanks for the feedback @JukkaL !  Addressed your comments & rebased.

**Comment by JukkaL** [comment]
Thanks for the updates! Looks good now.
